src/Symfony/Bridge/PhpAmqp/Connection.php

```diff
@@ -0,0 +1,192 @@
+
+/**
+ * An AMQP connection.
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class Connection
+{
+    private $amqpConnectionCredentials;
+    private $exchangeName;
+    private $queueName;
+    private $debug;
+
+    /**
+     * @var \AMQPChannel|null
+     */
+    private $amqpChannel;
+
+    /**
+     * @var \AMQPExchange|null
+     */
+    private $amqpExchange;
+
+    /**
+     * @var \AMQPQueue|null
+     */
+    private $amqpQueue;
+
+    public function __construct(array $amqpConnectionCredentials, string $exchangeName, string $queueName, bool $debug = false)
+    {
+        $this->amqpConnectionCredentials = $amqpConnectionCredentials;
+        $this->exchangeName = $exchangeName;
+        $this->queueName = $queueName;
+        $this->debug = $debug;
+    }
+
+    public static function fromDsn(string $dsn, bool $debug = false)
+    {
+        if (false === ($parsedUrl = parse_url($dsn))) {
+            throw new \InvalidArgumentException(sprintf('The given AMQP DSN "%s" is invalid.', $dsn));
+        }
+
+        $pathParts = explode(trim($parsedUrl['path'] ?? '', '/'), '/');
+
+        $amqpOptions = array(
+            'host' => $parsedUrl['host'] ?? 'localhost',
+            'port' => $parsedUrl['port'] ?? 5672,
+            'vhost' => $pathParts[0] ?? '/',
+            'queue_name' => $queueName = $pathParts[1] ?? 'messages',
+            'exchange_name' => $queueName,
+        );
+
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $parsedQuery);
+
+            $amqpOptions = array_merge($amqpOptions, $parsedQuery);
+        }
+
+        return new self($amqpOptions, $amqpOptions['exchange_name'], $amqpOptions['queue_name'], $debug);
+    }
+
+    /**
+     * @throws \AMQPException
+     */
+    public function publish(string $body, array $headers = array())
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->exchange()->publish($body, null, AMQP_NOPARAM, array(
+            'headers' => $headers,
+        ));
+    }
+
+    /**
+     * Waits and gets a message from the configured queue.
+     *
+     * @throws \AMQPException
+     */
+    public function waitAndGet(): ?\AMQPEnvelope
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $message = null;
+
+        try {
+            $this->queue()->consume(function (\AMQPEnvelope $envelope) use (&$message) {
+                $message = $envelope;
+
+                return false;
+            });
+        } catch (\AMQPQueueException $e) {
+            if (404 == $e->getCode()) {
+                // If we get a 404 for the queue, it means we need to setup the exchange & queue.
+                $this->setup();
+
+                return $this->waitAndGet();
+            }
+
+            throw $e;
+        }
+
+        return $message;
+    }
+
+    private function channel(): \AMQPChannel
+    {
+        if (null === $this->amqpChannel) {
```

Comment:
What about a guard clause instead to reduce the complexity?

```php
if (null !== $this->amqpChannel) {
    return $this->amqpChannel;
}

$connection = new \AMQPConnection($this->amqpConnectionCredentials);
if (false === $connection->connect()) {
    throw new \AMQPException('Could not connect to the AMQP server. Please verify the provided DSN.');
}

return $this->amqpChannel = new \AMQPChannel($connection);
```

I would do the same for all other methods in this file.